examples/knative/go/kn-go-tagging/tagging/tagging.go

```diff
@@ -0,0 +1,154 @@
+	SecretPath string `envconfig:"VCENTER_SECRET_PATH" default:""`
+	TagName    string `envconfig:"TAG_NAME" required:"true"`
+	TagAction  string `envconfig:"TAG_ACTION" default:"attach"`
+
+	DebugLogs bool `envconfig:"DEBUG" default:"true"`
+}
+
+// NewClient creates a client for vSphere, tagging, and CloudEvent methods.
+func NewClient(ctx context.Context) (*Client, error) {
+	var env EnvConfig
+	if err := envconfig.Process("", &env); err != nil {
+		return nil, fmt.Errorf("process environment variables: %w", err)
+	}
+
+	sc, err := newSOAPClient(ctx, env)
+	if err != nil {
+		return nil, fmt.Errorf("create vSphere SOAP client: %w", err)
+	}
+
+	rc, err := newRESTClient(ctx, sc.Client, env)
+	if err != nil {
+		return nil, fmt.Errorf("create vSphere REST client: %w", err)
+	}
+	tm := tags.NewManager(rc)
+
+	ce, err := client.NewHTTP(cehttp.WithPort(env.Port))
+	if err != nil {
+		return nil, fmt.Errorf("create CloudEvents client: %w", err)
+	}
+
+	return &Client{
+		soap:       sc,
+		rest:       rc,
+		cloudEvent: ce,
+		tagMgr:     tm,
+		envConfig:  env,
+	}, nil
+}
+
+// Run starts the CloudEvents receiver.
+func (c *Client) Run(ctx context.Context) error {
+	logging.FromContext(ctx).Debugw("start CloudEvents receiver", zap.Int("port", c.envConfig.Port))
+	return c.cloudEvent.StartReceiver(ctx, c.handler)
+}
+
+// Close releases resources used by the client.
+func (c *Client) Close(ctx context.Context) {
+	// Use fresh context to perform logout operations.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
+	defer cancel()
+
+	if err := c.soap.Logout(ctx); err != nil {
+		logging.FromContext(ctx).Warnw("log out from vSphere SOAP API", zap.Error(err))
+	}
+
+	if err := c.rest.Logout(ctx); err != nil {
+		logging.FromContext(ctx).Warnw("log out from vSphere REST API", zap.Error(err))
+	}
+}
+
+func (c *Client) handler(ctx context.Context, event event.Event) protocol.Result {
+	logging.FromContext(ctx).Debugw("received new event", zap.String("event", event.String()))
+
+	// Retrieve the Managed Object Reference from the event.
+	moRef, err := parseEventMoRef(event)
+	if err != nil {
+		return fmt.Errorf("parse CloudEvent: %w", err)
+	}
+
+	tag, err := c.tagMgr.GetTag(ctx, c.envConfig.TagName)
+	if err != nil {
+		return cehttp.NewResult(http.StatusBadRequest, "get tag %q: %w", c.envConfig.TagName, err)
+	}
+
+	switch c.envConfig.TagAction {
+	case tagActionAttach:
+		err = c.tagMgr.AttachTag(ctx, tag.ID, moRef)
+		if err != nil {
+			return fmt.Errorf("attach tagID %q to %q: %w", tag.ID, moRef.Value, err)
+		}
+		logging.FromContext(ctx).Debugw("attach tag to vm", zap.String("tag", tag.ID), zap.String("VM", moRef.Value))
+	case tagActionDetach:
+		err = c.tagMgr.DetachTag(ctx, tag.ID, moRef)
+		if err != nil {
+			return fmt.Errorf("attach tagID %q to %q: %w", tag.ID, moRef.Value, err)
+		}
+		logging.FromContext(ctx).Debugw("detach tag from vm", zap.String("tag", tag.ID), zap.String("VM", moRef.Value))
+	default:
```

Comment:
nit: this should be validation earlier, e.g. in `main.go` so you fail fast due to invalid config (instead of failing handler every time which pollutes logs)